lib/Support/JSON.cpp

```diff
@@ -0,0 +1,134 @@
+//===- Json.cpp - Json Utilities --------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Utilities for file system path handling, supplementing the ones from
+// llvm::sys::path.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Support/JSON.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/OperationSupport.h"
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+namespace json = llvm::json;
+
+using namespace circt;
+using mlir::UnitAttr;
+
+// NOLINTBEGIN(misc-no-recursion)
+LogicalResult circt::convertAttributeToJSON(llvm::json::OStream &json,
+                                            Attribute attr) {
+  return TypeSwitch<Attribute, LogicalResult>(attr)
+      .Case<DictionaryAttr>([&](auto attr) {
+        json.objectBegin();
+        for (auto subAttr : attr) {
+          json.attributeBegin(subAttr.getName());
+          if (failed(convertAttributeToJSON(json, subAttr.getValue())))
+            return failure();
+          json.attributeEnd();
+        }
+        json.objectEnd();
+        return success();
+      })
+      .Case<ArrayAttr>([&](auto attr) {
+        json.arrayBegin();
+        for (auto subAttr : attr)
+          if (failed(convertAttributeToJSON(json, subAttr)))
+            return failure();
+        json.arrayEnd();
+        return success();
+      })
+      .Case<BoolAttr, StringAttr>([&](auto attr) {
+        json.value(attr.getValue());
+        return success();
+      })
+      .Case<IntegerAttr>([&](auto attr) -> LogicalResult {
+        // If the integer can be accurately represented by a double, print
+        // it as an integer. Otherwise, convert it to an exact decimal string.
+        const auto &apint = attr.getValue();
+        if (apint.getActiveBits() >= std::numeric_limits<int64_t>::digits)
+          return failure();
+        json.value(apint.getSExtValue());
+        return success();
+      })
+      .Case<FloatAttr>([&](auto attr) -> LogicalResult {
+        const auto &apfloat = attr.getValue();
+        double value = apfloat.convertToDouble();
```

Comment:
Looks like this method /requires/ it fits in a double already-- change the unit test to `FloatType::getF128` instead of `getF64` and you get:

```
CIRCTSupportTests: /home/will/src/sifive/circt/llvm/llvm/lib/Support/APFloat.cpp:5427: double llvm::APFloat::convertToDouble() const: Assertion `getSemantics().isRepresentableBy(semIEEEdouble) && "Float semantics is not representable by IEEEdouble"' failed.
 #0 0x00000000003fed4b llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/will/src/sifive/circt/llvm/llvm/lib/Support/Unix/Signals.inc:602:13
 #1 0x00000000003fcfb2 llvm::sys::RunSignalHandlers() /home/will/src/sifive/circt/llvm/llvm/lib/Support/Signals.cpp:105:18
 #2 0x00000000003ff3fd SignalHandler(int) /home/will/src/sifive/circt/llvm/llvm/lib/Support/Unix/Signals.inc:0:3
 #3 0x00007f585a23dbf0 __restore_rt (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x3dbf0)
 #4 0x00007f585a28abc7 __pthread_kill_implementation (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x8abc7)
 #5 0x00007f585a23db46 gsignal (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x3db46)
 #6 0x00007f585a2284b5 abort (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x284b5)
 #7 0x00007f585a2283d9 _nl_load_domain.cold.0 (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x283d9)
 #8 0x00007f585a2367b6 (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x367b6)
 #9 0x00000000003b27e8 llvm::APFloat::convertToDouble() const /home/will/src/sifive/circt/llvm/llvm/lib/Support/APFloat.cpp:5431:3
#10 0x0000000000447787 mlir::LogicalResult circt::convertAttributeToJSON(llvm::json::OStream&, mlir::Attribute)::$_4::operator()<mlir::FloatAttr>(mlir::FloatAttr) const /home/will/src/sifive/circt/lib/Support/JSON.cpp:69:22
```

(line numbers may be a little off/misleading, been poking at this a bit)

---

Anyway looks like this isn't a great way to check if it'll round-trip, consider adding a test where it doesn't to check (perhaps the F128 case, but not sure if others).

Does this work for NaN / infinity / pos / negative zero ?